File: Auth/Users.App/RegistrationService.cs
using Microsoft.Extensions.Options;
using System.Data;
using System.Security.Cryptography;
using Users.App.Interface;
using Users.App.Settings;
using Users.Domain;

namespace Users.App
{
    public class RegistrationService(IUserRepo repo, IRegistrationEmailService emailService, IOptions<RegisterSettings> registerSettings)
    {
        private readonly RegisterSettings _registerSettings = registerSettings.Value;

        public async Task<bool> RegisterFederatedAsync(FederatedUser fedUser)
        {
            var user = await repo.FindByExternalIdAsync(fedUser.Id);
            if (user == null)
            {
                user = new User
                {
                    Id = Guid.NewGuid(),
                    ExternalId = fedUser.Id,
                    Issuer = fedUser.Issuer,
                    Email = fedUser.Email,
                    Username = fedUser.Username ?? fedUser.Email,
                    Name = fedUser.Name,
                    CreatedAt = DateTime.UtcNow,
                    Roles = "customer"
                };
                await repo.InsertAsync(user);
                return true;
            }
            user.Issuer = fedUser.Issuer;
            user.Email = fedUser.Email;
            user.Username = fedUser.Username ?? fedUser.Email;
            user.Name = fedUser.Name;
            await repo.UpdateAsync(user);
            return true;
        }

        public async Task<RegistrationResult> RegisterAsync(string email, string password)
        {
            // Check if user already exists
            var existingUser = await repo.FindByEmailAsync(email);
            if (existingUser != null && existingUser.EmailConfirmed)
            {
                return new RegistrationResult { Success = true };
            }
            var hashedPassword = PasswordHasher.Hash(password);
            if (existingUser != null && existingUser.EmailConfirmed == false)
            {
                existingUser.Password = hashedPassword;
                var token = GenerateConfirmationToken();
                existingUser.EmailConfirmationToken = token;
                await repo.UpdateAsync(existingUser);
                var confUrl = $"{_registerSettings.EmailConfirmUrl}?token={Uri.EscapeDataString(token)}";
                await emailService.SendRegistrationEmailAsync(existingUser, email, confUrl);

                return new RegistrationResult { Success = true };
            }
            
            var expiresAt = DateTimeOffset.UtcNow.AddHours(_registerSettings.EmailConfirmationTokenExpiresInHours);
            var confirmationToken = GenerateConfirmationToken();
            var newUser = new User
            {
                Id = Guid.NewGuid(),
                Email = email,
                Username = email,
                Password = hashedPassword,
                EmailConfirmed = false,
                EmailConfirmationToken = confirmationToken,
                EmailConfirmationTokenExpiresAt = expiresAt,
                Issuer = "local",
                CreatedAt = DateTimeOffset.UtcNow,
                Roles = "customer"
            };

            await repo.InsertAsync(newUser);

            var url = $"{_registerSettings.EmailConfirmUrl}?token={Uri.EscapeDataString(confirmationToken)}";
            await emailService.SendRegistrationEmailAsync(newUser, email, url);

            return new RegistrationResult { Success = true };
        }

        public async Task<bool> ResendRegistrationEmail(string email)
        {
            var existingUser = await repo.FindByEmailAsync(email);
            if(existingUser == null || existingUser.EmailConfirmed)
            {
                return false;
            }
            var confirmationToken = GenerateConfirmationToken();

            bool sent = await emailService.SendRegistrationEmailAsync(
                existingUser,
                email,
                $"{_registerSettings.EmailConfirmUrl}?token={Uri.EscapeDataString(confirmationToken)}");
            if (!sent)
            {
                return false;
            }

            var expiresAt = DateTimeOffset.UtcNow.AddHours(_registerSettings.EmailConfirmationTokenExpiresInHours);
            existingUser.EmailConfirmationToken = confirmationToken;
            existingUser.EmailConfirmationTokenExpiresAt = expiresAt;
            await repo.UpdateAsync(existingUser);
            return true;
        }

        public async Task<EmailConfirmationResult> ConfirmEmailAsync(string token)
        {
            if (string.IsNullOrWhiteSpace(token))
            {
                return new EmailConfirmationResult
                {
                    Success = false,
                    Error = "InvalidToken",
                    ErrorDescription = "Token is required."
                };
            }

            var user = await repo.FindByConfirmationTokenAsync(token);
            if (user == null)
            {
                return new EmailConfirmationResult
                {
                    Success = false,
                    Error = "InvalidToken",
                    ErrorDescription = "Invalid confirmation token."
                };
            }

            if (user.EmailConfirmed)
            {
                return new EmailConfirmationResult
                {
                    Success = false,
                    Error = "AlreadyConfirmed",
                    ErrorDescription = "Email has already been confirmed."
                };
            }

            if (user.EmailConfirmationTokenExpiresAt == null || 
                DateTimeOffset.UtcNow > user.EmailConfirmationTokenExpiresAt.Value)
            {
                return new EmailConfirmationResult
                {
                    Success = false,
                    Error = "TokenExpired",
                    ErrorDescription = "Confirmation token has expired."
                };
            }

            user.EmailConfirmed = true;
            user.EmailConfirmationToken = null;
            user.EmailConfirmationTokenExpiresAt = null;
            await repo.UpdateAsync(user);

            return new EmailConfirmationResult { Success = true };
        }

        private static string GenerateConfirmationToken()
        {
            var bytes = new byte[32];
            using var rng = RandomNumberGenerator.Create();
            rng.GetBytes(bytes);
            return Convert.ToBase64String(bytes);
        }
    }

    public class RegistrationResult
    {
        public bool Success { get; set; }
        public string? Error { get; set; }
        public string? ErrorDescription { get; set; }
    }

    public class EmailConfirmationResult
    {
        public bool Success { get; set; }
        public string? Error { get; set; }
        public string? ErrorDescription { get; set; }
    }
}

File: NodeAuth/src/services/RegistrationService.ts
import { randomBytes, randomUUID } from "crypto";

import { RegisterSettings } from "../config/register";
import { FederatedUser } from "../models/FederatedUser";
import { User } from "../models/User";
import { IUserRepo } from "../repos/interfaces/IUserRepo";
import { PasswordHasher } from "../utils/passwordHasher";
import { IRegistrationEmailService } from "./interfaces/IRegistrationEmailService";

export class RegistrationService {
  constructor(
    private readonly repo: IUserRepo,
    private readonly emailService: IRegistrationEmailService,
    private readonly settings: RegisterSettings,
  ) {}

  async registerFederatedAsync(fedUser: FederatedUser): Promise<boolean> {
    let user = await this.repo.findByExternalIdAsync(fedUser.id);
    if (!user) {
      user = {
        id: randomUUID(),
        externalId: fedUser.id,
        issuer: fedUser.issuer,
        email: fedUser.email,
        username: fedUser.username ?? fedUser.email,
        name: fedUser.name,
        createdAt: new Date(),
        roles: "customer",
        emailConfirmed: true,
        isLocked: false,
        failedLoginCount: 0,
      };
      await this.repo.insertAsync(user);
      return true;
    }

    user.issuer = fedUser.issuer;
    user.email = fedUser.email;
    user.username = fedUser.username ?? fedUser.email;
    user.name = fedUser.name;
    await this.repo.updateAsync(user);
    return true;
  }

  async registerAsync(email: string, password: string): Promise<RegistrationResult> {
    const existingUser = await this.repo.findByEmailAsync(email);
    if (existingUser && existingUser.emailConfirmed) {
      return { success: true };
    }

    if (existingUser && !existingUser.emailConfirmed) {
      existingUser.password = PasswordHasher.hash(password);
      const token = this.generateConfirmationToken();
      existingUser.emailConfirmationToken = token;
      existingUser.emailConfirmationTokenExpiresAt = this.addHours(
        new Date(),
        this.settings.emailConfirmationTokenExpiresInHours,
      );
      await this.repo.updateAsync(existingUser);
      const confirmUrl = `${this.settings.emailConfirmUrl}?token=${encodeURIComponent(
        token,
      )}`;
      await this.emailService.sendRegistrationEmailAsync(
        existingUser,
        email,
        confirmUrl,
      );
      return { success: true };
    }

    const confirmationToken = this.generateConfirmationToken();
    const newUser: User = {
      id: randomUUID(),
      email,
      username: email,
      password: PasswordHasher.hash(password),
      emailConfirmed: false,
      emailConfirmationToken: confirmationToken,
      emailConfirmationTokenExpiresAt: this.addHours(
        new Date(),
        this.settings.emailConfirmationTokenExpiresInHours,
      ),
      issuer: "local",
      createdAt: new Date(),
      roles: "customer",
      isLocked: false,
      failedLoginCount: 0,
    };

    await this.repo.insertAsync(newUser);

    const url = `${this.settings.emailConfirmUrl}?token=${encodeURIComponent(
      confirmationToken,
    )}`;
    await this.emailService.sendRegistrationEmailAsync(newUser, email, url);

    return { success: true };
  }

  async resendRegistrationEmail(email: string): Promise<boolean> {
    const existingUser = await this.repo.findByEmailAsync(email);
    if (!existingUser || existingUser.emailConfirmed) {
      return false;
    }

    const confirmationToken = this.generateConfirmationToken();
    const confirmUrl = `${this.settings.emailConfirmUrl}?token=${encodeURIComponent(
      confirmationToken,
    )}`;
    const sent = await this.emailService.sendRegistrationEmailAsync(
      existingUser,
      email,
      confirmUrl,
    );
    if (!sent) {
      return false;
    }

    existingUser.emailConfirmationToken = confirmationToken;
    existingUser.emailConfirmationTokenExpiresAt = this.addHours(
      new Date(),
      this.settings.emailConfirmationTokenExpiresInHours,
    );
    await this.repo.updateAsync(existingUser);
    return true;
  }

  async confirmEmailAsync(token: string): Promise<EmailConfirmationResult> {
    if (!token?.trim()) {
      return {
        success: false,
        error: "InvalidToken",
        errorDescription: "Token is required.",
      };
    }

    const user = await this.repo.findByConfirmationTokenAsync(token);
    if (!user) {
      return {
        success: false,
        error: "InvalidToken",
        errorDescription: "Invalid confirmation token.",
      };
    }

    if (user.emailConfirmed) {
      return {
        success: false,
        error: "AlreadyConfirmed",
        errorDescription: "Email has already been confirmed.",
      };
    }

    if (
      !user.emailConfirmationTokenExpiresAt ||
      Date.now() > user.emailConfirmationTokenExpiresAt.getTime()
    ) {
      return {
        success: false,
        error: "TokenExpired",
        errorDescription: "Confirmation token has expired.",
      };
    }

    user.emailConfirmed = true;
    user.emailConfirmationToken = undefined;
    user.emailConfirmationTokenExpiresAt = undefined;
    await this.repo.updateAsync(user);

    return { success: true };
  }

  private generateConfirmationToken(): string {
    return randomBytes(32).toString("base64");
  }

  private addHours(date: Date, hours: number): Date {
    return new Date(date.getTime() + hours * 60 * 60 * 1000);
  }
}

export interface RegistrationResult {
  success: boolean;
  error?: string;
  errorDescription?: string;
}

export interface EmailConfirmationResult {
  success: boolean;
  error?: string;
  errorDescription?: string;
}
